go-controller/pkg/util/kube.go

```diff
@@ -644,45 +688,123 @@ func GetLbEndpoints(slices []*discovery.EndpointSlice, svcPort kapi.ServicePort,
 				klog.V(5).Infof("Skipping FQDN slice %s/%s", slice.Namespace, slice.Name)
-			} else { // endpoints are here either IPv4 or IPv6
-				validSlices = append(validSlices, slice)
+				continue
+			case discovery.AddressTypeIPv4:
+				endpointsPointer = &v4EndpointsTmp
+			case discovery.AddressTypeIPv6:
+				endpointsPointer = &v6EndpointsTmp
+			}
+
+			if endpointsPointer == nil {
+				continue
+			}
+
+			for _, ep := range slice.Endpoints {
+				// readiness selection: include endpoints that are either ready
+				// or terminating and serving.
+				if !(includeAllEndpoints || IsEndpointReady(ep) || IsEndpointServing(ep)) {
+					continue
+				}
+
+				// add to v4 or v6 candidate endpoints
+				if len(ep.Addresses) > 0 {
+					*endpointsPointer = append(*endpointsPointer, ep)
+				}
 			}
 		}
 	}
-
-	serviceStr := ""
-	if service != nil {
-		serviceStr = fmt.Sprintf(" for service %s/%s", service.Namespace, service.Name)
+	if len(v4EndpointsTmp) > 0 {
+		out.V4Endpoints = removeDuplicateAddressesFromEndpoints(v4EndpointsTmp)
 	}
-	// separate IPv4 from IPv6 addresses for eligible endpoints
-	for _, endpoint := range getEligibleEndpoints(validSlices, service) {
-		for _, ip := range endpoint.Addresses {
-			if utilnet.IsIPv4String(ip) {
-				klog.V(5).Infof("Adding endpoint IPv4 address %s port %d%s",
-					ip, out.Port, serviceStr)
-				v4IPs.Insert(utilnet.ParseIPSloppy(ip).String())
+	if len(v6EndpointsTmp) > 0 {
+		out.V6Endpoints = removeDuplicateAddressesFromEndpoints(v6EndpointsTmp)
+	}
+
+	klog.V(5).Infof("LB candidate endpoints for %s/%s on port %d are: %v / %v ",
+		slices[0].Namespace, slices[0].Labels[discovery.LabelServiceName], out.Port,
+		out.V4Endpoints, out.V6Endpoints)
+
+	return out
+}
 
-			} else if utilnet.IsIPv6String(ip) {
-				klog.V(5).Infof("Adding endpoint IPv6 address %s port %d%s",
-					ip, out.Port, serviceStr)
-				v6IPs.Insert(utilnet.ParseIPSloppy(ip).String())
+func getEndpointInfoForLogging(ep discovery.Endpoint) string {
+	return fmt.Sprintf("node=%s, addresses=%s, conditions=Ready:%t, Serving:%t, Terminating:%t",
+		*ep.NodeName, ep.Addresses, *ep.Conditions.Ready, *ep.Conditions.Serving,
+		*ep.Conditions.Terminating)
+}
 
+// removeDuplicateAddressesFromEndpoints takes a list of endpoints and removes duplicate IP addresses
+// within the same endpoint and across endpoints. As discussed in  https://github.com/kubernetes/kubernetes/issues/121556
+// the API server does not guarantee unique IPs and it's up to the endpointslice API clients
+// to do the necessary deduplication. A conservative approach, as described in
+// https://github.com/kubernetes/kubernetes/issues/121231#issuecomment-1781547629 is applied:
+// we keep the IP address on the endpoint with the worse conditions (w.r.t ready, serving, terminating)
+// and remove the IP from the endpoint with the better conditions.
+func removeDuplicateAddressesFromEndpoints(endpoints []discovery.Endpoint) []discovery.Endpoint {
```

Comment:
I think it would work better as an overall strategy if you had a map of addresses to (selected) endpoint. Less iterations overall.

One iteration to build the map, as you do you zero out the endpoint address. Another iteration on the map, to populate the selected endpoint addresses.

Unless you require any kind of order stability, but that is not apparent otherwise I think it would be good to mention it.

Reply:
I'm totally with changing the logic, but I didn't get your approach :) 
Let's say I have an empty map of addresses to endpoints and I fill it in by iterating through all endpoints. The resulting values in the map will for sure include duplicate endpoints, since an endpoint can easily have more than one IP address. The main problem I faced is that the endpoints for an endpoint slice are not uniquely identified, it's just a list :-|

Reply:
I like this solution!

One advantage is that I am setting `endpoints[i].Addresses` to an empty slice at the end of each iteration through the endpoints, so I don't need to remove any element from any slice. 

The implementation with the map of addresses to endpoint pointers would have easier to read, but in unit tests we expect a certain order of endpoints and, within endpoints, a certain order of addresses :| 

So I kept the endpoint ordering by dealing with indices instead of pointers as you suggested, but I'm a bit afraid that the unpredictable order of addresses within endpoints will yield flakes in our unit tests :|